### ISSUE TITLE
Probing fails and what is probing acually trying to do?

### ISSUE THREAD

**Issue by nmbrg** [favonia/cloudflare-ddns#857]
The log says its trying to probe 1.1.1.1 and 1.0.0.1, but fails.

```
Probing 1.1.1.1 and 1.0.0.1 . . .
Both 1.1.1.1 and 1.0.0.1 are not working; sticking to 1.1.1.1 now.
The network might be temporarily down; will redo probing later.
```

first and foremost, ive got no clue on what it is actually trying to do (couldnt find this info anywhere)?
second, why might it fail? it isnt just one time or now and then, its every single time since i started the container, for several days.

**Comment by favonia** [comment]
@nmbrg Thanks for the question! And also thanks for using this tool. :grinning: I feel the proper solution is to add another entry in the FAQ section of README. But let me try to answer your questions here:

1. The probing is testing whether one can connect to Cloudflare's (not someone else's) 1.1.1.1 and 1.0.0.1 successfully. The probing is to work around some ISPs and routers which (incorrectly) intercept 1.1.1.1 or maybe 1.0.0.1. It will use whichever can connect to Cloudflare. The probing involves checking whether the HTTP response is signed by Cloudflare, but nothing more, and thus the network traffic overhead should be small (by using the HTTP HEAD method).

2. The failure could be that you restart the container when you reboot. Often your network need more time to be fully up. If the container tries to probe 1.1.1.1 and 1.0.0.1 while your network is not fully set up yet, the probing will not work. Also, because a Docker container is stateless (unless you mount something), it cannot remember the result of the previous run. I need more details about your machines/network/configurations to guess what might have caused the failure. On the other hand, the failure should not affect the functionality---the code is ready to handle all these errors.

**Comment by nmbrg** [comment]
@favonia ok, thanks for the explanation! When I come think about it I think it might be my adguard home servers that perhaps blocks those IP. I’ll try to see if that’s the case.

Will cloudflare-ddns work fine even tho it can’t talk to those IP (one of them)? Or do I need to make sure it can to be able to use cloudflare-ddns?

**Comment by favonia** [comment]
@nmbrg I see. Is it correct to say that the updating itself is not working, so you checked the logging, and then found the probing errors? If so, you can ignore the messages about the probing for now. The errors about the updating itself are perhaps more relevant. May I see the error messages about the updating and/or detecting?

Unfortunately, using services at these IPs (1.0.0.1 and 1.1.1.1) is still the most reliable ways to get IPv4 addresses from Cloudflare. Several other workarounds could sometimes get IPv6 instead IPv4 addresses. On the other hand, if you are fine with non-Cloudflare servers, you can use `url:...` to use another server to get the IPv4 address. There are plenty of such services on the internet.

**Comment by favonia** [comment]
I still want to see the error messages to check that it's not a bug about the code itself, if possible. Thanks!

**Comment by nmbrg** [comment]
> I still want to see the error messages to check that it's not a bug about the code itself, if possible. Thanks!

im not sure there are any errors. i just got stumped by the probing thing and why it did not manage to do the probing.

this is what the log say:
```
Checking the IP addresses in about 58s . . .
📖 Probing 1.1.1.1 and 1.0.0.1 . . .
   😞 Both 1.1.1.1 and 1.0.0.1 are not working; sticking to 1.1.1.1 now
   💡 The network might be temporarily down; will redo probing later
🌐 Detected the IPv4 address: xxx.xxx.xxx.xxx
🤷 The A records of "my.domain.tld" are already up to date (cached)
🤷 The A records of "my.domain.tld" are already up to date (cached)
🤷 The A records of "my.domain.tld" are already up to date (cached)
```

**Comment by favonia** [comment]
@nmbrg Hmm I now think maybe 1.1.1.1 is working but the imposed time limit (1 second) is too short. On the other hand I'm concerned about a longer time limit. Could you possibly help me understand your comment "every single time since i started the container"? Do you mean every time before "🌐 Detected the IPv4 address:"?

If so, I have a few new ideas, then: the probing should be skipped if you can detect the IP addresses without any problem. Would you still be concerned if it only appears 1-2 times? This does not solve all cases but should solve yours.

**Comment by favonia** [comment]
@nmbrg I opened a new issue #858 based on my current understanding. Please let me know if it matches your experience. I feel there might be other issues not covered by #858 so I will leave this issue open. :grin: At very least, I think the probing should be documented somewhere.

**Comment by nmbrg** [comment]
> @nmbrg Hmm I now think maybe 1.1.1.1 is working but the imposed time limit (1 second) is too short. On the other hand I'm concerned about a longer time limit. Could you possibly help me understand your comment "every single time since i started the container"? Do you mean every time before "🌐 Detected the IPv4 address:"?
> 
> If so, I have a few new ideas, then: the probing should be skipped if you can detect the IP addresses without any problem. Would it bother you less if it only appears 1-2 times? This does not solve all cases but should solve yours.

what i meant with that comment was that the container is always running, and im using `UPDATE_CRON=@every 1m`. Im by no means an expert, but afaik everytime that job runs it does the probing, and the logs are updated with the info i pasted earlier. so that was what i meant with "every single time i started the container", which in hindsight was really shitty explanation on when it happens, since the container is always running. :)

actually it doesnt bother me at all, if i know everything works fine regardless of the probing "error". tho i find it weird it does not manage to do the probing (ive got a 250mbit fiber connection so its not a slow connection), but if it works it works. :)

**Comment by favonia** [comment]
@nmbrg Thank you. The "cron" thing can indeed be very misleading. Usually it's a cron service running a "cron job" to update DNS records, but that's not the case here. There's no cron service, and the `UPDATE_CRON` is read and emulated by the updater itself. Throughout the execution there's only one program, the updater, running. This is how the (compressed) size of the image can be less than 5 MiB.

I am still very curious about the reason why there's probing error if the network is fast, and I want to improve the UI. To be precise, 250 Mib/s refers to the throughput, and it can still take a long time for your machine to reach the Cloudflare server (though I would still be surprised that the total time for the HTTP handshake can take more than 1 second given a fiber connection).

**Comment by favonia** [comment]
@nmbrg I guess the only way to figure out is to measure the actual time you need to complete the probing. If you can run `curl` on the same machine, the connection is roughly equivalent to these two:
```sh
curl --head https://1.1.1.1/
curl --head https://1.0.0.1/
```
If the `curl` commands also take a lot of time (> 1 second), then the network is "slow" for the purpose of quick probing. If the `curl` commands end instantly, there's some unknown bug to fix.

**Comment by nmbrg** [comment]
> @nmbrg I guess the only way to figure out is to measure the actual time you need to complete the probing. If you can run `curl` on the same machine, the probing is roughly equivalent to these two commands:
> 
> ```shell
> curl --head https://1.1.1.1/
> curl --head https://1.0.0.1/
> ```
> 
> If the `curl` commands also take a lot of time (> 1 second), then the network is "slow" for the purpose of quick probing. If the `curl` commands end instantly (< 0.5 seconds), then there's a bug to fix.

Not sure how to measure the time for those commands exactly, and they do not return any time for it to complete. But the moment i press enter they return all the lines with information instantaneously.

Do you need me to paste it here?

**Comment by favonia** [comment]
No I don't need to see the output. The following command will show the total connection times:
```
curl --output /dev/null --silent --write-out 'Total time: %{time_total}s\n' --head https://1.1.1.1/         
```
Just to double check, the Docker container is running on the same machine as the one running this `curl` command, right? It is really strange that `curl` finishes its job quickly while the Golang equivalent seems to have trouble.

**Comment by nmbrg** [comment]
yes, its beeing run on the same machine as the one running the container.

this is the output:
```
curl --output /dev/null --silent --write-out 'Total time: %{time_total}s\n' --head https://1.1.1.1/  
Total time: 0.062786s
```

i recently moved from a synology nas to unraid. And i can see that i had the same "issue" when it was running on the synology nas.

ran the same command on the synology nas and it gave me this:
```
curl --output /dev/null --silent --write-out 'Total time: %{time_total}s\n' --head https://1.1.1.1/
Total time: 0.191013s
```

**Comment by favonia** [comment]
@nmbrg Thank you. The time record of 0.06 seconds is certainly fast enough to finish any probing. I don't understand the root issue (that is, somehow the Go equivalent takes more than 1 second while it should be done in less than 0.1 seconds), but I plan to do the following fine-tuning:

1. Implement an even faster probing by just finishing the TLS handshake (before starting the HTTP request).
2. Recall the successful connection to 1.1.1.1 and skip the testing entirely.

Both should be helpful in your case. The first one further reduces the probing time and the second should apply to your situation (because you managed to detect a valid IP, a sign that no probing is needed). I will report when I have made these two changes. Thanks for your reporting and information for me to improve this tool! :heart:

**Comment by nmbrg** [comment]
awesome! will try when it gets pushed to a new version on latest tag. :)
